Model class for TensorFlow

ahtsan opened the pull request:
After effort on #481 to explore the possibility of constructing Garage Model based on `tf.keras.Model`, we decided to come up with our own Garage Model with the guidance from #507.

* A example of Gaussian MLP Model based on Garage Model
* Unit tests served as examples on different use cases.

### Design ###
* A model will have inputs, outputs and parameters.
* It needs to be built before using.
* Flexibility for custom serialization handling. There is a `BaseModel` which basically describes the interface, and `AutoPickableModel` which handles pickling automatically. Subclassed model will have to implement **how to get parameters/weights** and **gow to set parameters/weights.**

### Reuse ###
To reuse the parameters in a model, we build the same model with different input, where it can be a `tf.placeholder` or the output from another network.
* In the case of TensorFlow, we adopt a `parent-child` design. We have a parent model which contains the network structure/configuration, and everytime someone wants to use/reuse the model, a child model is spawned with the given input. We enforce the variable sharing by `tf.variable_scope` with a unique variable scope that belongs to the parent model.
* In the case of Keras, since variables sharing is done by calling a model with the input, we can simply wrap that operation under `model.build()`, which will look like
```python
def __init__(self):
  self.model = SomeCustomModel()
  ...

def build(self, *inputs):
  return self.model(*inputs)
```

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        pass
+
+
+class AutoPickableModel(BaseModel, metaclass=abc.ABCMeta):
```
i don't think this needs to be called autopickleable anymore. it can just be `Model`.

we may want to merge this with the `TfModel` class if we think that we won't have non-TF models in the future.

ahtsan replied:
In that case, `Model` by default will handle pickle and if someone wants to have another pickle implementation, he will inherit from `BaseModel`?

ryanjulian replied:
yes.

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+    parent model and can be accessed by calling model.child_2, where
+    child_2 is the name given when building this particular child model.
+    If a child is constructed without given a name, the name "default" will
+    be used.
```
* Note that each can only be built once
* Use the verb build/built, since "construct" already has a very specific meaning in Python

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+    (model.default.outputs)        model.child_2.outputs
+
+    The parameters inside a model will be initialized when calling build().
+    Do not call tf.global_variable_initializers() after building a model as it
```
I think this should be in its own paragraph and have some way of calling out it's importance

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        self._name = name or self._name
+        self._is_parent = True
+        self._is_built = False
+        self._variable_scope = tf.variable_scope(self.name, reuse=False)
```
did you figure out a way to detect duplicate names?

ahtsan replied:
TensorFlow will complain if two models have the same name, when one of them is built and another one tries to build.

ryanjulian replied:
does it raise an exception or only a warning?

ryanjulian replied:
okay it's fine for now. perhaps we can at least catch the exception in build and produce a more helpful error message (then rethrow it)

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        self._is_built = False
+        self._variable_scope = tf.variable_scope(self.name, reuse=False)
+        self._name_scope = tf.name_scope("default")
+        self._siblings = {}
```
are they siblings or are they children?

i think siblings/twins/clones is a more apt description. children implies a tree structure (and you will want to reserve that for referrring to hierarchical mdoels). you should make this consistent with the documentation above.

ahtsan replied:
I will change the naming accordingly.

ryanjulian replied:
is the resolution to call this `_graphs` or `_networks` instead then?

ahtsan replied:
We can separate into two classes: `Model` and `Networks`, and then have this called `_networks`.

ryanjulian replied:
okay, that sounds good.

ryanjulian replied:
i will resolve when you land the Network class

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+    @classmethod
+    def _new_child(cls, other, name):
+        """Class method for constructing child model with the given name."""
+        twin = cls(**other.init_spec[1])
```
is it a twin, a subling, or a clone? be consistent.

ahtsan replied:
It will be a clone of `Network`.

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        self._siblings = {}
+
+    @classmethod
+    def _new_child(cls, other, name):
```
maybe this should just be an overload of `__copy__`?

ahtsan replied:
This function will create a clone of the `Network`, so yes it can be an overload of `__copy__` in the `Network` class.

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+    @classmethod
+    def _new_child(cls, other, name):
+        """Class method for constructing child model with the given name."""
+        twin = cls(**other.init_spec[1])
```
please make init_spec typed. it can just be a plain data object or a namedtuple

ryanjulian replied:
not necessary if we use a different pickling strategy

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        """Class method for constructing child model with the given name."""
+        twin = cls(**other.init_spec[1])
+        twin._name = name
+        twin._is_parent = False
```
again it's important to get the terminology right.

parent/child implies a tree structure.

ryanjulian replied:
i think you will need different names

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        pass
+
+    @property
+    def variable_scope(self):
```
do these really need to be publicly accessible?

ryanjulian replied:
you didn't address this?

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        return self._variable_scope
+
+    @property
+    def name_scope(self):
```
does this really need to be publicly accessible?

ryanjulian replied:
you didn't address this?

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        return tf.get_default_session().run(self._get_variables())
+
+    @parameters.setter
+    def parameters(self, parameters):
```
You might want to take a page out of the `torch.nn.Module` book and add a strict/non-strict mode here

krzentner replied:
That sounds like something for a later pull request?

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+    def __getstate__(self):
+        """Object.__getstate__."""
+        if not self._is_parent:
+            raise ValueError("Child of a model should not be pickled!")
```
why not?

ahtsan replied:
Under the new design, only `Model` should be pickled.

ryanjulian replied:
can't you remove this in the new design, since networks will not be pickleable anyway?

ryanjulian commented on tests/garage/tf/models/test_new_model.py:
```diff
@@ -0,0 +1,77 @@
+        input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        model = GaussianMLPModel(output_dim=2)
+        model.build(input_var)
+        data = np.random.random((3, 5))
```
please don't use random data in tests. it makes them non-deterministic.

ryanjulian commented on tests/garage/tf/models/test_new_model.py:
```diff
@@ -0,0 +1,77 @@
+        input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        model = GaussianMLPModel(output_dim=2)
+        model.build(input_var, name="custom_model")
+        data = np.random.random((3, 5))
```
please don't use random data in tests. it makes them non-deterministic.

ryanjulian commented on tests/garage/tf/models/test_new_model.py:
```diff
@@ -0,0 +1,77 @@
+        model = GaussianMLPModel(output_dim=2)
+        model.build(input_var)
+        model.build(input_var, name="model_2")
+        data = np.random.random((3, 5))
```
please don't use random data in tests. it makes them non-deterministic.

ryanjulian commented on tests/garage/tf/models/test_new_model.py:
```diff
@@ -0,0 +1,77 @@
+        model = GaussianMLPModel(output_dim=2)
+        model.build(input_var, name="model_2")
+        model.build(input_var)
+        data = np.random.random((3, 5))
```
please don't use random data in tests. it makes them non-deterministic.

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        elif name in self._siblings:
+            raise ValueError("Subgraph {} already exists!".format(name))
+        elif not self._is_parent:
+            raise ValueError("Child of a model should not be built!")
```
i think you could make this structural rather than an exception by splitting the interns of the model into two classes.

* `Model` class which holds the variable scope
* Zero or more `Network` (or `Graph` or `SubGraph`) classes which use that variable scope.

Then you can have `build` create and add a graphs to the Model class. It is not possible to call `build` on a graph (only a model can do that).

ahtsan replied:
Yes this can also avoid circular reference.

ryanjulian commented on tests/garage/tf/models/test_new_model.py:
```diff
@@ -0,0 +1,77 @@
+from tests.fixtures import TfGraphTestCase
+
+
+class TestNewModel(TfGraphTestCase):
```
TestModel

ryanjulian commented on tests/garage/tf/models/test_new_model.py:
```diff
@@ -0,0 +1,77 @@
+import pickle
```
test_model.py

ryanjulian commented on tests/garage/tf/models/test_new_model.py:
```diff
@@ -0,0 +1,77 @@
+import numpy as np
+import tensorflow as tf
+
+from garage.tf.models.gaussian_mlp_model import GaussianMLPModel
```
you should implement your own simple model in this test, rather than conflating the test with GaussianMLPModel. unit tests should be self-contained.

ryanjulian commented on tests/garage/tf/models/test_new_model_pickle.py:
```diff
@@ -0,0 +1,63 @@
+from garage.tf.models.gaussian_mlp_model import GaussianMLPModel
+
+
+class TestNewModelPickling(unittest.TestCase):
```
This can be a different test case in `test_model.py`

ryanjulian replied:
We usually try to keep a 1:1 mapping between source units and test files.

ahtsan replied:
I am having issue when I called `reset_default_graph()`. I guess the unit tests are run in parallel and somehow when `reset_default_graph()` is called, it is in another `tf.Session` from another unit test.

CatherineSue replied:
Em……I don't think it is true that when you call `reset_default_graph` it is called in another session. This doesn't make sense. Why don't use TfGraphTest?

ahtsan replied:
When I run the above tests standalone, it is fine. When I put it under `test_new_model.py` and run, it said I am calling `reset_default_graph()` within a session, which I didn't.

CatherineSue replied:
Is test_new_model running under the TfGraphTest? Then there is a session created in the setUp.

ahtsan replied:
but in principle, each setUp() is called before each test? If the tests are done one by one, it should be fine.

CatherineSue replied:
I don't think it would be fine. `setUp` is called before the test is called. So there is a session created.

CatherineSue replied:
What do you mean by `put it under test_new_model.py`?

ahtsan replied:
I first tried to put the pickle test under `TfGraphTest` and I did not create new session. I just used the one created in setUp().

CatherineSue replied:
I think by `put it under test_new_model.py`, the class should become a test case under the TestModel class. In this case, if `tf.reset_default_graph()` is called then there will be an error because of attempting to reset graph when a session is active. 

I suppose you need to call `reset_default_graph()` because unpickle uses the same variable scope. Usually calling the function in unittest is unpredictable because multiple graphs may be created and it is hard to know which one is default.

There are multiple ways to do this without calling it. 
`with tf.Graph().as_default()` or `sess = tf.Session(tf.Graph())`

ahtsan replied:
Yes, you are right. `with tf.Session(graph=tf.Graph()) as sess:` does the job.

ryanjulian commented on tests/garage/tf/models/test_new_model.py:
```diff
@@ -0,0 +1,77 @@
+    def test_model_creation_with_custom_name(self):
+        input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        model = GaussianMLPModel(output_dim=2)
+        model.build(input_var, name="custom_model")
```
it doesn't seem to me that you ever interact with the outputs from `build` in these tests.

you also never test a hierarchical model (a model with a model as input to build), which i think is a key use case.

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+import tensorflow as tf
+
+
+class BaseModel(abc.ABC):
```
This class needs really comprehensive documentation on the input/output types for each of these methods and properties. do not skimp.

ryanjulian replied:
i like the docstrings. please document their argument/return types.

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+    def __setstate__(self, data):
+        """Object.__setstate__."""
+        # TODO(krzentner): There has got to be a better way.
+        out = type(self)(*data['args'], **data['kwargs'])
```
@krzentner

krzentner replied:
I don't really know of a better way of doing this. What really concerns me here isn't the use of `type()`, but that if the model implementer forgets a parameter in `init_spec()`, then we could silently lose data. Since this would be very hard to debug (and we promised not to lose data), this would be very bad.

ryanjulian replied:
hmm very good point.

i still think quick_init() is too magical though. 

can we use something like your builder class idea instead?

ryanjulian replied:
https://github.com/rlworkgroup/proposals/pull/3

krzentner replied:
Yeah, I have also clearly articulated why I don't like `quick_init()`, but this is worse. This allows losing data for the same reason as `Serializable`, but *also* if a field is forgotten in `init_params`.

What stands in the way of doing this:
 - Make `build` a factory method which produces a subgraph.
 - Make Model naively serializable (since it would basically just contain the same things as `init_params()` would return).

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+            raise ValueError("Child of a model should not be built!")
+        else:
+            twin = type(self)._new_child(self, name)
+            setattr(self, name, twin)
```
@krzentner is there a better way?

ryanjulian replied:
KR and I discussed this and agree that this `setattr()` strategy is not needed now. if users really want to access the network later, we can expose a property `networks` which is a dict of the networks.

e.g.
```python
m = MyModel()
bar = m.build(foo)
charlie = m.build(bob, name='alice')

assert m.networks['default'].output == bar
assert m.networks['default'].input == foo
assert m.networks['alice'].input == bob
assert m.networks['alice'].output == charlie

ryanjulian commented on garage/tf/models/gaussian_mlp_model.py:
```diff
@@ -0,0 +1,206 @@
+"""GaussianMLPModel."""
```
Can you should how this would be used in GaussianMLPPolicy?

CatherineSue replied:
Agreed, that would be great.

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+    def _get_variables(self):
+        return {
+            v.name: v
+            for v in tf.global_variables(
```
there is a new way of doing this in 1.12

https://www.tensorflow.org/api_docs/python/tf/VariableScope#get_collection
https://www.tensorflow.org/api_docs/python/tf/VariableScope#global_variables

CatherineSue commented on tests/garage/tf/models/test_new_model_pickle.py:
```diff
@@ -0,0 +1,63 @@
+            model_data = pickle.dumps(model)
+
+        sess.close()
+        tf.reset_default_graph()
```
1. Why reset default graph here?
2. I think this op is strange here because you don't know which graph is default and where it is created. Why don't you create the sess inside a graph?
with tf.Graph().as_default() something like this

ahtsan replied:
I want to remove all previous defined variables to illustrate pickling.

CatherineSue replied:
Thanks for the explanation. I replied this issue in another comment.

CatherineSue commented on garage/tf/models/gaussian_mlp_model.py:
```diff
@@ -0,0 +1,206 @@
+                 std_share_network=False,
+                 std_hidden_sizes=(32, 32),
+                 min_std=1e-6,
+                 max_std=None,
```
Please group `init_std, min_std, max_std` together, 
and `learn_std, adaptive_std, std_share_network` together. 
and `hidden_sizes, hidden_nonlinearity, output_nonlinearity` together 
and `std_hidden_sizes, std_hidden_nonlinearity, std_output_nonlinearity` together.

CatherineSue commented on garage/tf/models/gaussian_mlp_model.py:
```diff
@@ -0,0 +1,206 @@
+
+    def _build(self, *inputs):
+        action_dim = self._output_dim
+        state_input = inputs[0]
```
What could inputs[1] be?

ryanjulian replied:
inputs must be a tuple because there could be more than one.

i think this object should warn when there are too many inputs to unpack. perhaps we can make that a feature of the class `Model`

ahtsan replied:
e.g. Input can contain state and actions.

CatherineSue replied:
Thanks for the explanation. I agree inputs should be a tuple. But I think using `state_input=inputs[0]` is weird. i.e. If this is a multi-goal env, the inputs would be observation plus goals, then the state_input should also be observation plus goals. If this is a Q function model, then state_input should be observation plus action.

ryanjulian replied:
This is a difficult one to get right.

I chose `tuple` for the interface because it's universal and can use namedtuples if you want. However, using namedtuples for input would require that the model implementer provide you access to the type (e.g. as a class method). That's kind of ugly.

```python
class MyModel(TfModel):
    
    @classmethod
    def input_type(cls):
        return namedtuple('MyModelInputs', 'state', 'task')

    def _build(self, inputs):
        state = inputs.state

# ...
m = MyModel()
inputs = MyModel.input_type(state=my_state_var, task=my_task_var)  # this is a huge pain
out1, out2 = m.build(*inputs)

# also valid
m = MyModel()
inputs = MyModel()
out1, out2 = m.build(my_state_var, my_task_var)  # slight ambiguous at input, nice unpacking at the output
```

What if we make inputs/outputs a dict? (e.g. like kwargs).

That is a slightly more complicated interface than tuples but at least provides an opportunity for validation. 

I don't think it quite gets the dot-notation @CatherineSue wants, but it's close.

Unfortunately it makes unpacking really messy.

```python
class MyModel(TfModel)
    def build(self, **inputs):
        state_input = inputs['state']  # KeyError if it's missing
        # ...
        return dict(out1=out1, out2=out2)

# ...
m = MyModel()
outs = m.build(state=my_state_var, task=my_task_var)  # you can't unpack a dict because it's unordered
out1 = outs['out1']  # eww
out2 = outs['out2']
```

I think it's important to remember that kwargs can also be treated as positional arguments, which provides the solution here IMO

```python
class MyModel(TfModel):
    
    def _build(self, state=None, task=None):
        assert state, "You must provide a state input tensor"
        assert task, "You must provide a task input tensor"

        # ...
        return out1, out2
        
# ...
m = MyModel()
out1, out2 = m.build(state=my_state_var, task=my_task_var)

# also totally valid
m = MyModel()
out1, out2 = m.build(my_state_var, my_task_var)

# AssertionError
m = MyModel()
out1, out2 = m.build(my_state_var)

# AssertionError
m = MyModel()
out1, out2 = m.build(state=my_state_va)
```

ahtsan replied:
Then the interface would have `def _build(self)`? We would have to specify very carefully in docstring that inputs are expected, but the inputs depend on the Model. It can be `state`, or both `state` and `action`.

ahtsan replied:
or it will still be `def _build(self, *inputs)` and we will pass actual kwargs for each models

ryanjulian replied:
No the interface would be unchanged.

```python
class BaseModel(abc.ABC):

    def build(self, *inputs):
        pass


class Model(BaseModel, metaclass=abc.ABCMeta):
    # ...


class TfModel(Model, metaclass=abc.ABCMeta):

    def build(self, *inputs, name=Nome):
        # ...

    def _build(self, *inputs, name=None):
        pass

class MyModel(TfModel):

    def _build(self, state, task, name=None):
        # ...
        return dist.sample(), mean, std



# Works fine
m = MyModel()
a, mean, std = m.build(my_state_var, my_task_var)

# Also works fine
m = MyModel()
a, mean, std = m.build(state=my_state_var, task=my_task_var, name='Foo')

# Also works fine
m = MyModel()
a, mean, std = m.build(state=my_state_var, task=my_task_var, 'Foo')

# Also works fine
m = MyModel()
a, mean, std = m.build(task=my_task_var, state=my_state_var, 'Foo')

# Error
m = MyModel()
a, mean, std = m.build(my_state_var, name='Foo')

# Error
m = MyModel()
a, mean, std = m.build(task=my_task_var)

# etc.
```

CatherineSue replied:
I agree with @ryanjulian 's proposal. Another question maybe related to the outputs unpack here: what's the purpose of these fileds [here](https://github.com/rlworkgroup/garage/pull/512/files/96e6512f704cb687464f55823b68c5306a74d760#diff-f27d474d4aed442478e7e44fdc2de671R201)? I don't see where it's explicitly used.

ahtsan replied:
I tried to make the outputs to be attributes of the `Model`.

ryanjulian replied:
the user should only be able to access the outputs of a model through the return value of `build` and through the `outputs` property of a net.

We can try it this way first. If it turns out that it makes for a lot of boilerplate code, we can come back and try something different with more magic.

```python
m = MyModel()
foo = m.build(bar)
assert m.output == foo  # AttributeError
assert m.networks['default'].output == foo
baz = m.build(bap, name='alice')
assert m.output == baz  # AttributeError
assert m.networks['alice'] == baz
```

CatherineSue commented on garage/tf/models/gaussian_mlp_model.py:
```diff
@@ -0,0 +1,206 @@
+    def _build(self, *inputs):
+        action_dim = self._output_dim
+        state_input = inputs[0]
+        with tf.variable_scope("dist_params"):
```
```  suggestion
        with tf.variable_scope('dist_params'):
```

ryanjulian replied:
you did not address the comment

CatherineSue commented on garage/tf/models/gaussian_mlp_model.py:
```diff
@@ -0,0 +1,206 @@
+                    hidden_nonlinearity=self._hidden_nonlinearity,
+                    output_nonlinearity=self._output_nonlinearity,
+                    output_b_init=b,
+                    name="mean_std_network")
```
'mean_std_network'
I don't want to create multiple comments about this, please replace double quotation to single quotation. They are appeared in here and the unittests.

ryanjulian replied:
please don't resolve comments until the resolved code shows on GitHub

krzentner commented on garage/tf/models/gaussian_mlp_model.py:
```diff
@@ -0,0 +1,206 @@
+            raise NotImplementedError
+
+    @property
+    def init_spec(self):
```
Again, I think this method shouldn't exist. We would be better off with something like `Serializable.quick_init()` then this.

CatherineSue commented on garage/tf/models/gaussian_mlp_model.py:
```diff
@@ -0,0 +1,206 @@
+        Positional and named arguments (args and kwargs) which, when passed
+        to `__init__`, will construct this model.
+        """
+        return (), {
```
What would positional args be?

ryanjulian replied:
in order to provide all args and all kwargs, an init_spec must be `(args, kwargs) --> tuple[list, dict]`, even if `args` is an empty list.

this is why i say above that we should create a type for init_spec.

I think @krzentner has a slightly more sophisticated proposal which achieves the same thing.

https://github.com/rlworkgroup/proposals/pull/3

krzentner commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        self._default_parameters = None
+
+    @property
+    def init_spec(self):
```
I think we should replace this function with something similar to `quick_init()`. This will prevent us from accidentally losing user data in the future, which is something we promise not to do.

ryanjulian replied:
i would prefer something less magical like your proposal here https://github.com/rlworkgroup/proposals/pull/3

krzentner replied:
I would also be perfectly happy with that. The important thing is that we don't set ourselves up to shoot feet (our own or otherwise).

ryanjulian replied:
i'm going to resolve all but one comment which addressing pickling strategy

CatherineSue commented on tests/garage/tf/models/test_new_model_pickle.py:
```diff
@@ -0,0 +1,63 @@
+                model_pickled.outputs[:2], feed_dict={input_var: data})
+
+            # result[0] is sample(), which is supposed to be different
+            assert np.array_equal(results[1:], results2[1:])
```
Why would the sample be different if the global seed is the same?

ahtsan replied:
Since it is not inheriting from `TfGraphTestCase` and I forgot to call `ext.set_seed()`.

CatherineSue commented on tests/garage/tf/models/test_new_model_pickle.py:
```diff
@@ -0,0 +1,63 @@
+            model_pickled.build(input_var)
+
+            results2 = sess.run(
+                model_pickled.outputs[:2], feed_dict={input_var: data})
```
I talked about this in the old model [pr](https://github.com/rlworkgroup/garage/pull/481#discussion_r252490100), I think making this a namedtuple is better. It is less required if a user often calls outputs instead of part of it.

ryanjulian replied:
i agree. can we make that a feature of `Model`?

ahtsan replied:
or a feature of `Network`? So that we have `model.network_1.mean_var`, etc.

ryanjulian replied:
a feature of network

ryanjulian commented on garage/tf/models/gaussian_mlp_model.py:
```diff
@@ -0,0 +1,206 @@
+            'std_parameterization': self._std_parameterization,
+        }
+
+    def _build(self, *inputs):
```
use actual kwargs. the * notation varargs is only for the inferface

```python
def _build(self, state=None):
    assert state, "You must provide a state tensor input"
```

ryanjulian replied:
use actual arg names here

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+        # TODO(krzentner): There has got to be a better way.
+        out = type(self)(*data['args'], **data['kwargs'])
+        out._default_parameters = data['parameters']
+        self.__dict__.update(out.__dict__)
```
the coverage system seems to think this hasn't been tested.

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+                        tf.get_default_session().run(
+                            tf.variables_initializer(variables))
+                        if self._default_parameters:
+                            self.parameters = self._default_parameters
```
coverage says this is not tested

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+                            self.parameters = self._default_parameters
+                        self._is_built = True
+        elif name in self._siblings:
+            raise ValueError("Subgraph {} already exists!".format(name))
```
coverage says this is not tested

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+            if name in parameters:
+                var.load(parameters[name])
+            else:
+                warnings.warn('No value provided for variable {}'.format(name))
```
Coverage says this is not tested

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,313 @@
+    If a child is constructed without given a name, the name "default" will
+    be used.
+
+    The design is illustrated as the following:
```
@krzentner has some concerns about this design, as expressed in https://github.com/rlworkgroup/garage/pull/507#discussion_r256672737

I am posting it here so we can close the other PR.

One alternative is for `build` to return a Graph object (with inputs and outputs), rather than storing the Graph internally and providing implicit access via setattr.

I don't think we can get around subgraph duplication, since TensorFlow basically forces that design. We will just have to choose how to represent it.

krzentner replied:
Wow, PR's with shared commits are quite confusing.

I do think the alternative you describe is better, although I don't really understand the circumstances under which I would want to make multiple subgraphs from the same model. Can't I just make multiple model instances? Or does that not work?

ahtsan replied:
From my understanding, when you make multiple model instances, they will contain their own variables, bounded by their own `variable_scope`. Each of them will essentially be a subgraph. And then if we want those models to share the same variables, that's what this PR is trying to achieve, i.e. share `variable_scope`.

One example is DDPG, where the policy outputs an action, and passed to the Qfunction (QF) for optimizing policy loss. At the same time, samples from environment are passed to the same QF for optimizing QF loss. In this case, we have two subgraphs (policy loss and qf loss) but we share the same QF.

ryanjulian replied:
i think i was able to explain this to KR last night

ryanjulian commented:
I think this is developed-enough that we should push it forward in its current state (taking into account comments) and see where it takes us. This is our third stab at it and I think we won't get it any better until we just try it.

ahtsan commented:
Major changes:
- Fix most of the comments
- Add GaussianMLPPolicy with GaussianMLPModel. Examples are in `tests.garage.tf.policies.test_gaussian_mlp_policy_with_model`
- `Network` in `Model` are now namedtuple
- We don't need magic for pickling now because `Model` is picklable
- Use `tfp.distributions`, e.g. `tfp.distributions.MultivariateNormalDiag`

Discussons:
- Perhaps `BaseModel` would not be needed? `TfModel` basically is having different design with them. Also `Model` can be merged with `TfModel`?
- `tfp.distributions.kl_divergence` doesn't yield the same values on two same distributions as the existing method. Also, there is a flag `allow_nan_stats`. Consider `kl_divergence(A, B)` and `allow_nan_stats=True`, it will give `nan` when results are undefined, e.g. if there is some zero probabilities in `A` and `inf` if there is some zero probabilities in `B`. If `allow_non_stats=False`, an exception is raised. It looks like we have to handle these cases.

CatherineSue commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,173 @@
+                    obs_ph, name='log_prob_obs') - policy4.model.networks[
+                        'default'].distribution.log_prob(obs_ph)
+
+            likelihood_ratio_sym2 = tf.exp(log_prob_diff)
```
Why `tf.exp`?

ahtsan replied:
I was referring to the `likelihood_ratio_sym` in garage.
```python
    def likelihood_ratio_sym(self,
                             x_var,
                             old_dist_info_vars,
                             new_dist_info_vars,
                             name=None):
        with tf.name_scope(name, "likelihood_ratio_sym",
                           [x_var, old_dist_info_vars, new_dist_info_vars]):
            logli_new = self.log_likelihood_sym(x_var, new_dist_info_vars)
            logli_old = self.log_likelihood_sym(x_var, old_dist_info_vars)
            return tf.exp(logli_new - logli_old)
```

CatherineSue replied:
Isn't this function in tfp is the [log_likelihood](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Distribution#log_prob)?

CatherineSue replied:
You should add this interface in policy. This is also a trivial issue.

ahtsan commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,368 @@
+        pass
+
+    @property
+    def output(self):
```
Yes, I have the same thought. That's why I suggested `BaseModel` would not be needed.

krzentner replied:
Honestly, the only part of this API which seems reasonably common across Tensorflow and PyTorch is something like `output = model.build(input)`. PyTorch would usually call that method `forward`, but adding an alias is reasonable.

Having `input` and `output` fields is also unusual / considered bad practice in PyTorch, since that makes evaluating the `Module` multiple times weird.

ahtsan commented on garage/tf/models/gaussian_mlp_model.py:
```diff
@@ -0,0 +1,181 @@
+
+        rnd = tf.random.normal(
+            shape=mean_var.get_shape().as_list()[1:], seed=ext.get_seed())
+        action_var = rnd * tf.exp(std_var) + mean_var
```
`GaussianMLPModel` uses the existing distribution, i.e. `garage.tf.distributions.DiagonalGaussian`. It has `sample(self, dist_info)` but not `sample().`
`GaussianMLPModel2` uses `tfp.distribution.MultivariateNormalDiag`, which has `sample()`.
I wanted to show the difference so there are two GaussianMLPModels.

ryanjulian replied:
i think this is fine for now, since this is a work-in-progress

ahtsan commented on garage/tf/policies/gaussian_mlp_policy_with_model2.py:
```diff
@@ -0,0 +1,125 @@
+            name = 'get_reparam_action_sym'
+        new_dist_info_vars = self.dist_info_sym(obs_var, name=name)
+        new_mean_var, new_log_std_var = new_dist_info_vars[
+            'mean'], new_dist_info_vars['log_std']
```
Can we do `tf.log(dist.stddev)`?

ahtsan replied:
or we can make that a field in the namedtuple, so we have `model.networks['default'].log_std`.

ahtsan commented on garage/tf/policies/gaussian_mlp_policy_with_model2.py:
```diff
@@ -0,0 +1,125 @@
+        log_stds = np.vstack(
+            [path['agent_infos']['log_std'] for path in paths])
+        logger.record_tabular('{}/AverageStd'.format(self.name),
+                              np.mean(np.exp(log_stds)))
```
output of `tfp.distributions.kl_divergence(A, B)` is not the same as the output of `dist.kl_sym(A, B)`

ahtsan replied:
I think I found the reason. I was passing the log std when I construct the `tfp.distribution.MultivariateNormalDiag`. I will fix that.

ahtsan commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,173 @@
+
+        # * New way *
+        # equvaient to distribution.kl_sym()
+        kl_diff_sym2 = tfp.distributions.kl_divergence(
```
We can, just I didn't.

CatherineSue replied:
We should use policy.dist.kl_divergence(old_dist)

CatherineSue commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,173 @@
+        # likelihood_ratio_sym
+        # * Existing way *
+        likelihood_ratio_sym1 = policy1.distribution.likelihood_ratio_sym(
+            obs_ph, dist1_sym, dist2_sym, name="li_ratio_sym1")
```
single quotation

CatherineSue commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,134 @@
+
+    def __init__(self,
+                 env_spec,
+                 name="GaussianMLPPolicy",
```
single quotation

ahtsan commented:
@CatherineSue Please check the unit test and see if it makes more sense now.

CatherineSue commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,134 @@
+            name = "get_reparam_action_sym"
+        new_dist_info_vars = self.dist_info_sym(obs_var, name=name)
+        new_mean_var, new_log_std_var = new_dist_info_vars[
+            "mean"], new_dist_info_vars["log_std"]
```
single quotation

CatherineSue replied:
If this file will be deleted eventually, you can ignore the single quotation.

CatherineSue commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,134 @@
+    def __init__(self,
+                 env_spec,
+                 name="GaussianMLPPolicy",
+                 dist=DiagonalGaussian,
```
Is this GaussianMLPPolicyWithModel necessary? Could we just replace policy with model and tfp.distributions together?

ahtsan replied:
I just wanted to show the difference. Yes, It is not needed.
I wanted to show 
- difference after adding `GaussianMLPModel` by `GaussianMLPPolicy` and `GaussianMLPPolicyWithModel`
- difference after using `tfp.distributions` by `GaussianMLPolicyWithModel` and `GaussianMLPolicyWithModel2`

ryanjulian replied:
we can commit both for now and then gradually delete one as we continue refactoring

CatherineSue commented on garage/tf/policies/gaussian_mlp_policy_with_model2.py:
```diff
@@ -0,0 +1,125 @@
+        :return:
+        """
+        if not name:
+            name = 'get_reparam_action_sym'
```
Why not 
``` python
def get_reparam_action_sym(self,
                               obs_var,
                               action_var,
                               old_dist_info_vars,
                               name='get_reparam_action_sym'):
```

ahtsan replied:
right

CatherineSue commented on garage/tf/policies/gaussian_mlp_policy_with_model2.py:
```diff
@@ -0,0 +1,125 @@
+    def __init__(self,
+                 env_spec,
+                 name='GaussianMLPPolicy',
+                 dist=tfp.distributions.MultivariateNormalDiag,
```
We shouldn't let users pass dist here. It is confusing.

ahtsan replied:
Can we assume that a GaussianMLPPolicy always uses MultivariateNormalDiag?

ryanjulian replied:
yes

CatherineSue commented on garage/tf/policies/gaussian_mlp_policy_with_model2.py:
```diff
@@ -0,0 +1,125 @@
+                 env_spec,
+                 name='GaussianMLPPolicy',
+                 dist=tfp.distributions.MultivariateNormalDiag,
+                 *args,
```
Could you specific args here, please? I want to find where [these params](https://github.com/rlworkgroup/garage/pull/512/files#diff-ee08069b2a03df62f0e02fcbbbf10d5cR97) are going to pass, but it's really hard to. 
Also this is the class user will be dealing direct with, this will make it very difficult to be sure what are the args and kwargs.

ahtsan replied:
Yes that will be more readable.

CatherineSue commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,181 @@
+            likelihood_ratio2 = self.sess.run(  # result
+                likelihood_ratio_sym2,
+                feed_dict={
+                    policy3.model.networks['default'].input: obs,
```
Is there a more concise way to call the inputs and outputs? It is exhausting to get the output in a simple policy by `policy.model.networks['default'].output`. 
This seems trivial if all the reviewers think it is ok then you can ignore this.

CatherineSue commented on tests/garage/tf/models/test_model.py:
```diff
@@ -0,0 +1,178 @@
+    def test_model_creation(self):
+        input_var = tf.placeholder(tf.float32, shape=(None, 5))
+        model = SimpleModel(output_dim=2)
+        outputs = model.build(input_var)
```
where is `outputs` used?

ahtsan replied:
in the `assert`

CatherineSue replied:
I still can’t see which assert. The two asserts here use out and model_out

ahtsan replied:
`output` are the output Tensors. It's in the `feed_dict`. I am trying to show the outputs of `model.build()` are actually same as `model.networks['default'].outputs`. Users can do either way to access it.

CatherineSue replied:
I guess my question is, can we pass outputs as the fetches in sess.run?

CatherineSue replied:
I c.

CatherineSue commented on tests/garage/tf/models/test_model.py:
```diff
@@ -0,0 +1,178 @@
+
+    def _build(self, obs_input):
+        h1, _ = self._simple_model_1.build(obs_input)
+        return self._simple_model_2.build(h1)[1]
```
Could you show me the computation graph of this ComplicatedModel?

ahtsan replied:
<img width="810" alt="screen shot 2019-02-18 at 9 24 33 pm" src="https://user-images.githubusercontent.com/10270515/52991950-aae0a880-33c3-11e9-88cf-cf5df6bcabc4.png">

CatherineSue commented on tests/garage/tf/models/test_model.py:
```diff
@@ -0,0 +1,178 @@
+
+        assert np.array_equal(results, results2)
+
+    def test_model_in_model_pickle_same_parameters(self):
```
this looks like `test_complicated_model_pickle` to me. where are the `same_parameters`?

CatherineSue commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,181 @@
+and P4. All four policies are implemented with GaussianMLPModel.
+P1 and P2 are from GaussianMLPPolicyWithModel, which uses
+garage.tf.distributions while P3 and P4 from GaussianMLPPolicyWithModel2,
+which uses tf.distributions.
```
tfp.distributions

CatherineSue commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,181 @@
+
+        # * New way *
+        # this will be tf.distributions
+        policy3.dist_info_sym(obs_ph, name='p3_sym')
```
does this have a return? where is the return used?
will this func be used in GaussianMLPPolicyWithModel2? if it is not, we can discard this interface.

ahtsan replied:
Yes, this will be used. This will actually call `model.build()` with the given placeholder. We can rename it to be more clear, e.g. `policy3.build_dist(obs_ph, name="p3_sym")`.

CatherineSue commented on garage/tf/models/gaussian_mlp_model2.py:
```diff
@@ -0,0 +1,178 @@
+
+        action_var = distribution.sample(seed=ext.get_seed())
+
+        return action_var, log_std_var, distribution
```
why not return `mean_var`?

ahtsan replied:
we can get `mean_var` by `distribution.mean()`.

ryanjulian replied:
this will be easier to make right once we get rid of garage.distributions

CatherineSue commented on garage/tf/models/gaussian_mlp_model2.py:
```diff
@@ -0,0 +1,178 @@
+
+    def network_output_spec(self):
+        """Network output spec."""
+        return ['sample', 'log_std', 'distribution']
```
please rename `distribution` to `dist`. abbr is more convenient here.

ahtsan replied:
Ok

CatherineSue commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,181 @@
+        # equvaient to distribution.log_likelihood_sym(X, dist)
+        log_prob_sym2 = policy3.model.networks[
+            'default'].distribution.log_prob(
+                policy3.model.networks['default'].input)
```
why the value is `input`? shouldn't the value be the sampled action?

ahtsan replied:
This is just illustration on how `log_likelihood_sym` will look like.

CatherineSue commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,181 @@
+                    policy_pickled.model.networks['default'].input: data
+                })
+
+        assert np.array_equal(outputs, outputs2)
```
Are we abandoning `tensor_utils.compile_function()`? In this test, all the new ways use `sess.run`.

ahtsan replied:
I am not sure. That was discussed in another PR. Using `sess.run` is just my personal preference.

ryanjulian replied:
i have no preference. do whatever seems best for now and we can fix this style in a different PR

ahtsan commented:
This is the computational graphs for `test_gaussian_mlp_policy_with_model`.
<img width="1024" alt="screen shot 2019-02-18 at 9 31 41 pm" src="https://user-images.githubusercontent.com/10270515/52992259-cbf5c900-33c4-11e9-9769-b71a76a696e7.png">

krzentner commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,368 @@
+    def __getstate__(self):
+        """Object.__getstate__."""
+        new_dict = self.__dict__.copy()
+        del new_dict['_networks']
```
Wait, so `Model` knows about the `_networks` field, but doesn't actually have it? That's weird. We should just merge `TfModel` into `Model`, since it's in the `tf` directory already.

ryanjulian replied:
agreed

ahtsan replied:
We have to remove the `networks` when pickling. The networks are connectivity and I think they can be removed.

krzentner commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,368 @@
+    |    | (Network) |/               \|(Network) | |
+    |    =============                 ============ |
+    |       |                            |          |
+    =================================================
```
This diagram is great. Please make sure you keep it.

ahtsan commented:
Thanks for the comments.

More changes:
* Fix comments.
* Remove `input` / `output` / `inputs` / `output` and merged `TfModel` with `Model`.
* Add `likelihood_ratio_sym` interface in gaussian mlp policy.

We will keep either `GaussianMLPPolicyWithModel` or `GaussianMLPPolicyWithModel2`(I prefer this).

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,296 @@
+        model = SimpleModel()
+        # To use a model, first create a placeholder.
+        # In the case of TensorFlow, we create a tf.placeholder.
+        input_ph = tf.placeholder(tf.float32)
```
it this actually valid tensorflow?

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,296 @@
+            c = namedtuple(network_name,
+                           [*spec, 'input', 'output', 'inputs', 'outputs'])
+            if isinstance(network.outputs, tuple):
+                assert len(spec) == len(network.outputs),\
```
please split with () per PEP8

ryanjulian commented on garage/tf/models/gaussian_mlp_model.py:
```diff
@@ -0,0 +1,181 @@
+        """Network output spec."""
+        return ['sample', 'mean', 'std', 'std_param', 'dist']
+
+    def _build(self, state_input=None, dist=None):
```
according to your _build interface, _build implementations may have any number of positional args (*inputs) but no kwargs.

I think this code will function equivalently if you do
```python
def _build(self, state_input, dist):
   # ...
```
If the user forgets to include one or both of these, Python should raise an error which is equivalent to your `assert`s

ryanjulian commented on garage/tf/models/gaussian_mlp_model2.py:
```diff
@@ -0,0 +1,186 @@
+    def _build(self,
+               state_input=None,
+               dist=None,
+               hidden_nonlinearity=tf.nn.tanh,
```
all of these nonlinearities should be constructor params only.

I realize they are not pickleable. You can make them pickleable by using strings for the arguments and dispatching on those strings. I think keras has a utility for this, or alernatively just make your own.

```python
def tf_from_string(string):
   if string == 'tanh':
        return tf.nn.tanh
    elif string == 'tf.nn.softmax':
        return tf.nn.softmax
    # ...
    else:
        raise NotImplementedError
```

We can eventually make this part of the pickling code (or have the pickler automatically replace things like tf.nn.tanh in the pickled dict)

ahtsan replied:
They turn out to be picklable. I will put them in construct params.

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,158 @@
+            std_parameterization=std_parameterization)
+
+        with tf.variable_scope(name, reuse=False):
+            _, self._mean_var, self._log_std_var, _, self._dist = \
```
please split with () per PEP8

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,158 @@
+        new_action_var = new_mean_var + epsilon_var * tf.exp(new_log_std_var)
+        return new_action_var
+
+    def log_diagnostics(self, paths):
```
you can remove this -- it never gets called

ahtsan replied:
I saw it is called in `OffPolicyRLAlgorithm.log_diagnostics`.
```python
    def log_diagnostics(self, paths):
        """Log diagnostic information on current paths."""
        self.policy.log_diagnostics(paths)
        self.qf.log_diagnostics(paths)
```

ryanjulian replied:
(double check)

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model2.py:
```diff
@@ -0,0 +1,153 @@
+            feed_dict={self.model.networks['default'].input: observations})
+        return actions
+
+    def get_reparam_action_sym(self,
```
is this function called anywhere? if not, just deleted it.

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model2.py:
```diff
@@ -0,0 +1,153 @@
+        new_action_var = new_mean_var + epsilon_var * tf.exp(new_log_std_var)
+        return new_action_var
+
+    def log_diagnostics(self, paths):
```
if this isn't called anywhere, feel free to delete it

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,158 @@
+    def __init__(self,
+                 env_spec,
+                 name='GaussianMLPPolicy',
+                 dist=DiagonalGaussian,
```
by definition, doesn't a GaussianMLPPolicy use a DiagonalGaussian? I'm confused about why this is a parameter at all.

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,158 @@
+        actions = rnd * np.exp(log_stds) + means
+        return actions, dict(mean=means, log_std=log_stds)
+
+    def get_reparam_action_sym(self,
```
if this is not called anywhere, please delete it

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model2.py:
```diff
@@ -0,0 +1,153 @@
+        # There are two options:
+        # 1. We can enforce the user to pass the input and build the model
+        # once manually.
+        # 2. We can build it by default.
```
for now let's build it by default. eventually perhaps GaussianMLPPolicy will also have a `build()` interface.

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model2.py:
```diff
@@ -0,0 +1,153 @@
+        """Vectorized or not."""
+        return True
+
+    def likelihood_ratio_sym(self, obs_var, other_dist):
```
this assumes that `obs_var == state_input` from the constructor (or at least has the same shape), right?

ahtsan replied:
Yes

ryanjulian replied:
if you want to make that assumption you should test it. use an assert or an if+raise ValueError

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,296 @@
+
+    def __init__(self, model, inputs, name):
+        self._inputs = inputs
+        with tf.name_scope(name=name):
```
if we do this part inside model, you can remove the constructor dependence on model and not pass `self` (which is an anti-pattern).

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,296 @@
+        return self._outputs
+
+
+class TfModel(BaseModel):
```
As mentioned by @krzentner i think it is safe to merge `TfModel` and `Model` (into just `Model`) for now

ryanjulian commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,173 @@
+        policy3.dist_info_sym(obs_ph, name='p3_sym')
+        policy4.dist_info_sym(obs_ph, name='p4_sym')
+
+        ############################################################
```
all of these comment blocks can be different test cases.

you can uses `setUp()`/`tearDown()` to make sure all your common code is run before the test executs.

ryanjulian commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,173 @@
+                input_ph, tfp.distributions.MultivariateNormalDiag)
+
+            outputs2 = sess.run(
+                policy_pickled.model.networks['default'].sample,
```
seems like this is ripe for a helper function on `model`, but that can be for another PR

ryanjulian commented:
It looks like you need to add `tensorflow-probability` to setup.py

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,297 @@
+
+        Args:
+          inputs: Tensor input(s), recommended to be position arguments, e.g.
+            def build(self, state_input=None, action_input=None, name=None).
```
your example uses keyword arguments (e.g. `build(self, state_input=None, action_input=None, name=None)`), not positional arguments (e.g. `build(self, state_input, action_input, name=None)`)

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,297 @@
+        return self._name
+
+    def _get_variables(self):
+        return {v.name: v for v in tf.get_variable_scope().global_variables()}
```
don't use `tf.get_variable_scope()`, which points to the current scope in this thread regardless of what scope the model uses. add a test case to provide it to yourself.

instead, store the `Scope` object and then call `self._variable_scope.global_variables()`

ahtsan replied:
a `variable_scope` doesn't have `global_variables()`, only a `VariableScope` does, but we can only enter a `variable_scope`, not `VariableScope`.
I will use `tf.global_variables(self._variable_scope._name_or_scope)`.

ryanjulian replied:
ok

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,297 @@
+        new_dict['_default_parameters'] = self.parameters
+        return new_dict
+
+    def __setstate__(self, dict):
```
don't use dict as an argument, since it's a Python builtin. I'm surprised this wasn't caught by flake8.

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,121 @@
+                      state_info_vars=None,
+                      name='dist_info_sym'):
+        """Symbolic graph of the distribution."""
+        with tf.variable_scope(
```
i think this line should be removed and the scopes should be handled by `self.model.build()`

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,121 @@
+            std_output_nonlinearity=std_output_nonlinearity,
+            std_parameterization=std_parameterization)
+
+        with tf.variable_scope(name, reuse=False):
```
scopes should be handled 100% by `self.model`

ahtsan replied:
That's right!

ryanjulian commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,121 @@
+            std_output_nonlinearity=std_output_nonlinearity,
+            std_parameterization=std_parameterization)
+
+        with tf.variable_scope(name, reuse=False):
```
how do these scopes interact with the model scopes? is there a reason we even have them?

if you are going to use a scope repeatedly in an object, create it once and then save it (e.g. `self._variable_scope`)

ahtsan replied:
I agree the scope should be entirely handled by the model. I didn't remove them when I copy from `gaussian_mlp_policy`.

ryanjulian replied:
try it and see if it works. i think eventually GaussianMLPPolicy will inherit from a model (e.g. `class GaussianMLPPolicy(GaussianMLPModel)`) but we will need to do some more refactors before that works.

ryanjulian commented on tests/garage/tf/models/test_model.py:
```diff
@@ -0,0 +1,216 @@
+        return action
+
+
+class ComplicatedModel(Model):
```
what happens if one of the constructor params is a model?

CatherineSue replied:
Why mark this resolved? I’m also curious to know if it is possible to pass a simple model to a complicated model.

ryanjulian replied:
sorry i resolved it and then asked the same question in a more relevant line below.

please resolve this yourself when you're satisfied :)

ahtsan replied:
no worry I will add a test for this

ryanjulian commented on tests/garage/tf/models/test_model.py:
```diff
@@ -0,0 +1,216 @@
+                                             })  # noqa: E126
+        assert np.array_equal(results_1, results_2)
+
+    def test_model_in_model(self):
```
what if i use Foo as a constructor parameter to model Bar?

ryanjulian replied:
@CatherineSue see here

ryanjulian commented on tests/garage/tf/models/test_model.py:
```diff
@@ -0,0 +1,216 @@
+
+        assert np.array_equal(out, model_out)
+
+    def test_model_pickle(self):
```
did you address KR's concerns about the pickling procedure?

krzentner replied:
This looks right to me. Anson and I discussed it at some length.

ryanjulian commented:
Check out your coverage report and try to add a few tests for your missing lines to get the coverage higher.

I think a few more tests would get your to 90%

Don't worry about testing trivial things (like the interface methods, which just `pass`) or properties which just return a member variable.

ryanjulian commented on tests/garage/tf/policies/test_gaussian_mlp_policy_with_model.py:
```diff
@@ -0,0 +1,165 @@
+"""
```
can you add test which uses `GaussianMLPPolicyWithModel` in an algorithm?

ryanjulian commented on garage/tf/models/base.py:
```diff
@@ -0,0 +1,297 @@
+        network_name = name or 'default'
+
+        if not self._networks:
+            _variable_scope = tf.variable_scope(self._name, reuse=False)
```
store variable_scope rather than recreating it every time

ahtsan replied:
It will make `Model` unpicklable as `tf.variable_scope` is not picklable. But we can remove this field in `__getstate__` and construct it in `__setstate__`.

ahtsan commented:
### Changes:
* Assert likelihood_ratio_sym input shape same with model input
* Policy now handles scope by model
* Test on model as constructor param to another model
* Test on ppo with new GaussianMLPPolicyWithModel

### One modification on ppo:

`policy.dist_info_sym` with same name is called in both `_build_entropy_term` and `_build_policy_loss`, but they are actually accessing the same underlying variables.
However, inside `GaussianMLPPolicyWithModel`, we cannot call `dist_info_sym` with same name because of the design (we make sure 1-to-1 correspondence between `network` and tf.graph). Complying with the API, I changed the name in the `dist_info_sym` call in `_build_policy_loss`. We can indeed retrieve the target subgraph by `policy.model.networks['default'].dist`.

I also added `get_params` in `GaussianMLPPolicyWithModel` as it is required in optimizer. It is basically same as `model._get_variables()`, but getting the trainable variables. We can add an extra interface here.

### Notes:
I also copied the `Policy` and `StochasticPolicy` base class to be `Policy2` and `StochasticPolicy2`, but without inheriting from Parameterized.

It seems like `GaussianMLPPolicyWithModel`, which uses `garage.tf.distributions.DiagonalGaussian`, can easily replace existing GaussianMLPPolicy. If we want to use `GaussianMLPPolicyWithModel2`, which uses tfp.distributions.MultivariateNormDiag, we may need some more refactors in the algorithm.

Also, TensorFlow is amazing. `variable_scope` is not picklable but `VariableScope` is. So I store the `VariableScope` then and don't have to care about it when pickling.

CatherineSue commented on garage/tf/algos/npo.py:
```diff
@@ -406,7 +406,7 @@ def _build_entropy_term(self, i):
                     i.flat.obs_var,
                     i.flat.policy_state_info_vars,
-                    name="policy_dist_info_flat")
+                    name="policy_dist_info_flat_entropy")
```
Why do you change the name? And the name is inappropriate.

ahtsan replied:
I explained above. Basically the model design does not allow us to call `dist_info_sym` with the same name twice.

ahtsan replied:
Is it actually pointing to the same `policy_dist_info_flat` as in `_build_policy_loss`?

CatherineSue commented on garage/tf/policies/gaussian_mlp_policy_with_model.py:
```diff
@@ -3,10 +3,11 @@
 
 
-class GaussianMLPPolicyWithModel:
+class GaussianMLPPolicyWithModel(StochasticPolicy2):
```
Can we just abandon GaussianMLPPolicyWithModel for good? Here it inherits StochasticPolicy2. And GaussianMLPPolicyWithModel2 inherits from StochasticPolicy2, too? It is really confusing to me which to inherit from which.

ahtsan replied:
It will take some effort to make the existing `GaussianMLPPolicyWithModel2` work with algorithm because the API is very different.
`GaussianMLPPolicyWithModel2` does not inherit from `StochasticPolicy2`. The reason why `GaussianMLPPolicyWithModel` does is that it needs to comply with the algorithm API.

CatherineSue replied:
Yes it needs take the effort to refactor the algorithm, that’s what we have to do eventually. But it would make more sense if you delete GaussianMLPPolicyWithModel, have another ppoWithModel to compatible with the current GaussianMLPPolicyWithModel2.

CatherineSue replied:
IMO if you have some xx and xx2 naming, model should inherit policy, model2 should inherit policy2. But right now it doesn’t do this. It makes me confused.

ahtsan replied:
That makes sense. If everyone is ok with that, I will remove `GaussianMLPPolicyWithModel` and make a ppoWithModel.

ahtsan replied:
and I will name everything with model with suffix `WithModel`.

ahtsan replied:
with original `ppo` and `GaussianMLPPolicyWithModel`, most of the changes are about *How GaussianMLPolicy* will look like. I didn't expect to introduce `ppoWithModel` in this PR. That's why I want to make minimal changes to `ppo` but to show how `GaussianMLPolicy` can be used with an algorithm.

ahtsan replied:
If that makes more confusion, I agree it's better to move on and make `ppoWithModel`.

CatherineSue replied:
With your latest change, would PPO work with other policies that is not implemented with Model? I.e. GaussianLSTMPolicy.

CatherineSue replied:
I’m ok if we delete GaussianMLPPolicyWithModel and only add GaussianMLPPoliyWithModel2 in this pr. Add PPOWithModel seems to be the scope of another pr

ahtsan replied:
I can make `ppoWithModel` in another PR, so we won't be seeing test which uses `GaussianMLPPolicyWithModel` in an algorithm, in this PR.

zhanpenghe replied:
I agree with @ahtsan to make minimal changes to algorithms in this pr but I don't agree with writing `PPOWithModel`. We should just refactor algorithms when we are refactoring for using tf distributions. 

I will suggest the following: refactor policies/baselines with models and keep garage distribution at this point, then replace all distribution with tf distribution in the next pr.

I don't understand why model should inherit policy. @CatherineSue
